src/core/experimental/refcount.d

```diff
@@ -0,0 +1,545 @@
+    }
+
+    @nogc nothrow pure @trusted scope
+    this(this Q)(int) const
+    {
+        CounterType* support = cast(CounterType*) pureAllocate(2 * CounterType.sizeof);
+        static if (is(Q == immutable))
+        {
+            *support = 0;
+            rc = cast(immutable CounterType*) support;
+        }
+        else
+        {
+            *(support + 1) = 0;
+            rc = cast(CounterType*) (support + 1);
+        }
+        addRef();
+    }
+
+    private enum copyCtorIncRef = q{
+        rc = rhs.rc;
+        assert(rc == rhs.rc);
+        if (rhs.isInitialized())
+        {
+            assert(isShared() == rhs.isShared());
+            addRef();
+        }
+    };
+
+    @nogc nothrow pure @safe scope
+    this(return scope ref typeof(this) rhs)
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    // { Get a const obj
+    @nogc nothrow pure @safe scope
+    this(return scope ref typeof(this) rhs) const
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    @nogc nothrow pure @safe scope
+    this(return scope const ref typeof(this) rhs) const
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    @nogc nothrow pure @safe scope
+    this(return scope immutable ref typeof(this) rhs) const
+    {
+        mixin(copyCtorIncRef);
+    }
+    // } Get a const obj
+
+    // { Get an immutable obj
+    @nogc nothrow pure @trusted scope
+    this(return scope ref typeof(this) rhs) immutable
+    {
+        // Can't have an immutable ref to a mutable. Create a new RC
+        CounterType* support = cast(CounterType*) pureAllocate(2 * CounterType.sizeof);
+        *support = 0;
+        rc = cast(immutable CounterType*) support;
+        addRef();
+    }
+
+    @nogc nothrow pure @trusted scope
+    this(return scope const ref typeof(this) rhs) immutable
+    {
+        if (rhs.isShared())
+        {
+            // By implementation, only immutable RC is shared, so it's ok to inc ref
+            rc = cast(immutable) rhs.rc;
+            if (isInitialized())
+            {
+                addRef();
+            }
+        }
+        else
+        {
+            // Can't have an immutable ref to a mutable. Create a new RC
+            CounterType* support = cast(CounterType*) pureAllocate(2 * CounterType.sizeof);
+            *support = 0;
+            rc = cast(immutable CounterType*) support;
+            addRef();
+        }
+    }
+
+    @nogc nothrow pure @safe scope
+    this(return scope immutable ref typeof(this) rhs) immutable
+    {
+        mixin(copyCtorIncRef);
+    }
+    // } Get an immutable obj
+
+    @nogc nothrow pure @safe scope
+    ref __RefCount opAssign(return scope ref typeof(this) rhs)
+    {
+        if (rhs.isInitialized() && rc == rhs.rc)
+        {
+            return rhs;
+            //return this;
+        }
+        if (rhs.isInitialized())
+        {
+            rhs.addRef();
+        }
+        if (isInitialized())
+        {
+            delRef();
+        }
+        () @trusted { rc = rhs.rc; }();
```

Comment:
This is because `rhs` is `scope` and `this` is not. In hindsight, `rhs` probably should not be `scope` but it seems to be required by the L362 in the unittest `a2 = a3;`. Removing `scope` from `opAssign`'s `rhs` declaration yields the following error
```
refcount.d(362): Error: scope variable a3 assigned to non-scope parameter rhs calling core.experimental.refcount.__RefCount.opAssign
```

Maybe I should remove `scope` from the function attributes list?

Reply:
Removing scope doesn't work, but I'm no export on scope/return.

Reply:
>What happens when structs like TestRC are cast between mutable/shared/immutable?

Good food for thought. E.g. what if someone uses the `assumeUnique` idiom with an object that has a `__RefCount` as a member?

Reply:
My take: should be transparent to casting, i.e. work as things have worked until now, which means (for this and other reasons) that `__RefCount` must always use interlocked operations :(.